llvm/include/llvm/SandboxIR/SandboxIR.h

```diff
@@ -1386,6 +1390,26 @@ class PtrToIntInst final : public CastInst {
 #endif // NDEBUG
 };
 
+class BitCastInst : public CastInst {
+public:
+  static Value *create(Value *Src, Type *DestTy, BBIterator WhereIt,
+                       BasicBlock *WhereBB, Context &Ctx,
+                       const Twine &Name = "");
+  static Value *create(Value *Src, Type *DestTy, Instruction *InsertBefore,
+                       Context &Ctx, const Twine &Name = "");
+  static Value *create(Value *Src, Type *DestTy, BasicBlock *InsertAtEnd,
+                       Context &Ctx, const Twine &Name = "");
+
+  static bool classof(const Value *From) {
+    return isa<Instruction>(From) &&
```

Comment:
I believe isa + cast is forbidden.
```
if (Instruction *I = dyn_cast<Instruction>(From))
  return I->getOpcode() == Instruction::Opcode::BitCast;
return false;
```

Reply:
"Note that you should not use an isa<> test followed by a cast<>, for that use the dyn_cast<> operator."

Reply:
`isa` + `cast` is pretty common for these exact cases where you want to early return if the instruction is of the wrong type. Just grep for it in `llvm/lib`, you will find lots of examples that look exactly like this.

The alternative would look like:
```
auto *I = dyn_cast<Instruction>(From);
return I == nullptr ? false : I->getOpcode() == Instruction::Opcode::BitCast;
```
which is not as elegant and probably slower because of the extra control flow within `dyn_cast`.

> "Note that you should not use an isa<> test followed by a cast<>, for that use the dyn_cast<> operator."

I think this is meant to discourage things like:
```
  Foo *I = nullptr;
  if (isa<Foo>(V))
     I = cast<Foo>(V);
```

Reply:
I think both ways will probably optimize to the same code given they're doing the same thing.

can you do `return isa<BitCastInst>(From)`?

Reply:
Well it may work for now, but I don't think it is super safe. Once we introduce multi-Instructions then the class of `From->Val` won't necessarily be representative of the whole sandboxir::Value. So I think it is safer to keep it as is.

Reply:
but this isn't a multi-instruction? other multi-instructions will have their own implementation of `classof` that will have to take into account multiple LLVM instructions?

Reply:
The `LLVM Programmer’s Manual` says that you should use dyn_cast. I would follow that.

Reply:
It's not, but `From` could be in a multi-Instruction. And it is theoretically possible to have two separate multi-instructions that contain `llvm::BitCastInst`s as their `Val` values. Which is why I think that using `From->Val`'s class is not the right approach. It is safer to get the SandboxIR instruction that points to `From->Val` and get it's opcode.

Reply:
oh I see, I keep getting confused between LLVM IR and SandboxIR, I haven't spent long enough staring at SandboxIR code to get the namespaces straight in my head

I do slightly prefer tschuett's original suggestion over the current version, it's a bit less redundant even though it's an extra line

Reply:
OK I will update it to the `dyn_cast` one.